### PR TITLE
Iox #1196 fix lockfree queue warnings

### PR DESCRIPTION
**Pull request by MatthiasKillat**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Partial #1196

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -84,6 +90,16 @@ hicpp-*,
...
 #        void bar(Foo& foo) {...}
+#
+# * rule: cppcoreguidelines-non-private-member-variables-in-classes
+#   justification: Sometimes it makes sense to have protected members to extend the base class.
+#                  Even public members are useful in lightweight classes (effectively structs).
+#                  This unnecessarily limits design options.
+#
+# * rule: cppcoreguidelines-avoid-c-arrays
+# * rule: hicpp-avoid-c-arrays
+#   justicfication: C arrays are useful in low level code for buffers etc.
```
Comment (MatthiasKillat) [inline .clang-tidy:101]: @elBoberido If we do not want to suppress this globally I will add a suppression comment where it applies.

Note that I think using C arrays should be allowed and would be in favor of suppression.

File: iceoryx_hoofs/test/moduletests/test_lockfree_queue.cpp [hunk 2]
```diff
@@ -41,8 +42,11 @@ struct Integer
     int value{0};
 
     // so that it behaves like an int for comparison purposes
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) required for typed tests
     operator int() const
     {
+        /// @todo suppression does not work but it is a false positive
+        // NOLINTNEXTLINE(clang-analyzer-core.unitialized.UndefReturn) false positive - defaults to 0
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_lockfree_queue.cpp:49]: @elBoberido 

This suppression does not quite work and the warning is false positive or at least not clear to me. The return value should be well defined if the object is constructed (and we do not call it before).

File: iceoryx_hoofs/test/moduletests/test_lockfree_queue_buffer.cpp [hunk 3]
```diff
@@ -32,22 +32,11 @@ class LockFreeQueueBufferTest : public ::testing::Test
...
-    void SetUp()
-    {
-    }
-
-    void TearDown()
-    {
-    }
+    ~LockFreeQueueBufferTest() override = default;
 
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init) false positive
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_lockfree_queue_buffer.cpp:39]: @elBoberido This suppression also does not work for me and again I think it is a false positive. The buffer is default initialized in the default ctor, there should be no problem (at least the way I understand the warning).

The problem is that the rule may find valid problems with uninitialized members as well, but if it catches these kind of false positives it is useless and should be disabled I think.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.inl [hunk 4]
```diff
@@ -14,6 +14,11 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
+#ifndef IOX_HOOFS_CONCURRENT_LOCKFREE_QUEUE_BUFFER_INL
+#define IOX_HOOFS_CONCURRENT_LOCKFREE_QUEUE_BUFFER_INL
+
+#include "iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.hpp"
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.inl:20]: @elBoberido 
This is necessary to provide definitions locally in this file (but not for the way it is included). If not included, this causes clang-tidy warnings as it does not know about the context of inclusion.

This pseudo-cyclical inclusion (prevented by preprocessor) is weird design though.

File: .clang-tidy [hunk 5]
```diff
@@ -25,6 +25,11 @@ hicpp-*,
 
 -readability-identifier-naming,
 -readability-use-anyofallof,
+-readability-named-parameter,
+
+-cppcoreguidelines-non-private-member-variables-in-classes,
+
+-hicpp-named-parameter,
```
Comment (elBoberido) [inline .clang-tidy:32]: Could you move this up so that they are alphabetically sorted?

File: .clang-tidy [hunk 6]
```diff
@@ -88,7 +93,17 @@ hicpp-*,
...
+# * rule: cppcoreguidelines-non-private-member-variables-in-classes
+#   justification: Sometimes it makes sense to have protected members to extend the base class.
+#                  Even public members are useful in lightweight classes (effectively structs).
+#                  This unnecessarily limits design options.
+#
+# * rule: readability-named-parameter
+#   justicfication: For tag types which just guide overloading no name is needed.
+#                   Declaration of private functions does not require a name (no doxygen comment).
+#                   For callbacks omitting the name is less verbose than to cast an unsued parameter to void.
+#
```
Comment (elBoberido) [inline .clang-tidy:106]: Could you please add a small example which shows the violation and in the `readability-named-parameter` how we want it, e.g.
```
bad
  auto callable = [] (auto /*foo*/) {...};
good
  auto callable = [] (auto) {};
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.hpp [hunk 7]
```diff
@@ -51,6 +51,7 @@ class Buffer
   private:
     using byte_t = uint8_t;
 
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
     alignas(ElementType) byte_t m_buffer[Capacity * sizeof(ElementType)];
 
     ElementType* toPtr(index_t index) const noexcept;
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.hpp:57]: Please add a justification like `Required for low-level building block and wrapped in the abstraction`

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.inl (hunk 4 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/buffer.inl:20]: yes, it's required for clang-tidy but also other tools require it to cleanly parse the file

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp [hunk 8]
```diff
@@ -116,6 +116,7 @@ class IndexQueue
     ///    the default atomic constructor does not call the default constructor of the
     ///    underlying class.
     ///    See, http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0883r0.pdf
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
     Cell m_cells[Capacity];
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp:120]: Couldn't the `buffer` be used for this? Not in this PR but just for a future refactoring?

File: iceoryx_hoofs/test/moduletests/test_index_queue.cpp [hunk 9]
```diff
@@ -32,21 +32,9 @@ class IndexQueueTest : public ::testing::Test
...
+    IndexQueueTest() = default;
 
-    void SetUp()
-    {
-    }
-
-    void TearDown()
-    {
-    }
+    ~IndexQueueTest() override = default;
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_index_queue.cpp:37]: Didn't know that `override` also applies to dtors

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp [hunk 10]
```diff
@@ -29,14 +29,16 @@ using namespace ::testing;
...
 struct Integer
 {
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) required for typed tests
     Integer(uint64_t value = 0)
         : value(value)
     {
     }
 
-    uint64_t value{0};
+    uint64_t value = 0;
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:38]: Out of curiosity, did clang-tidy complain about this or is it just personal preference?

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp [hunk 11]
```diff
@@ -481,7 +471,8 @@ TYPED_TEST(ResizeableLockFreeQueueTest, DecreaseCapacityOfAPartiallyFilledQueue)
 
     uint64_t element = 0U;
     while (q.tryPush(element++))
-        ;
+    {
+    }
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:475]: We should have a clang-format which creates `while (cond) {}` in a single line

File: .clang-tidy (hunk 5 above)
Comment (MatthiasKillat) [reply .clang-tidy:32]: I was thinking about sorting them by category/prefix as first key.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 8 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp:120]: Better use some initialized array here as we want the ctor to be called for each element. Maybe even std::array.

File: iceoryx_hoofs/test/moduletests/test_index_queue.cpp (hunk 9 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_index_queue.cpp:37]: Apparently. It is in principle also a function and here the base implementation is virtual so clang-tidy complains to be explicit about overriding (even for default).

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp (hunk 10 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:38]: I do not remember. Will check later. Personally I can live with either syntax for primitive types but prefer `=` as it is closer to other languages.

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp (hunk 11 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:475]: Or, better, allow the previous code which is not bad to read and it is perfectly fine to have a loop and some small amount of logic in the loop condition. Textbook example would be quicksort (splitting by pivot).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp [hunk 12]
```diff
@@ -116,6 +116,7 @@ class IndexQueue
     ///    the default atomic constructor does not call the default constructor of the
     ///    underlying class.
     ///    See, http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0883r0.pdf
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp:119]: Please add a justification (and maybe remove the white spaces at the beginning of the comments above).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl [hunk 13]
```diff
@@ -14,12 +15,18 @@
...
+
+#include "iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp"
+
 namespace iox
 {
 namespace concurrent
 {
 template <uint64_t Capacity, typename ValueType>
-IndexQueue<Capacity, ValueType>::IndexQueue(ConstructEmpty_t) noexcept
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl:28]:
```suggestion
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl [hunk 14]
```diff
@@ -30,6 +37,7 @@ IndexQueue<Capacity, ValueType>::IndexQueue(ConstructEmpty_t) noexcept
 }
 
 template <uint64_t Capacity, typename ValueType>
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl:40]:
```suggestion
```

File: .clang-tidy [hunk 15]
```diff
@@ -88,7 +93,17 @@ hicpp-*,
...
 # * rule: readability-use-anyofallof
 #   justification: requires C++20 and std::ranges but we only use C++14
-
+#
+# * rule: cppcoreguidelines-non-private-member-variables-in-classes
+#   justification: Sometimes it makes sense to have protected members to extend the base class.
+#                  Even public members are useful in lightweight classes (effectively structs).
+#                  This unnecessarily limits design options.
+#
+# * rule: readability-named-parameter
```
Comment (FerdinandSpitzschnueffler) [inline .clang-tidy:102]:
```suggestion
# * rule: readability-named-parameter, hicpp-named-parameter
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 8 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp:120]: Changed to `std::array` (which does not allocate nor throw if used correctly (only in bounds checked access).

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp (hunk 10 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:38]: @elBoberido It is not required and hence undone. I believe clang-tidy complained when my language server was not working correctly for some reason but cannot reproduce it.

In my opinion both are fine for primitive types.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 8 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp]: @MatthiasKillat oh, then this was a misunderstanding on my side. I assumed `Cell` to be an alias to something like `uint8_t[N]`. The C array would have been okay here.

File: iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp (hunk 11 above)
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_resizeable_lockfree_queue.cpp:475]: No, not in this but maybe in a later PR.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl [hunk 16]
```diff
@@ -226,9 +232,9 @@ bool IndexQueue<Capacity, ValueType>::popIfFull(ValueType& index) noexcept
 }
 
 template <uint64_t Capacity, typename ValueType>
-bool IndexQueue<Capacity, ValueType>::popIfSizeIsAtLeast(const uint64_t requiredSize, ValueType& index) noexcept
+bool IndexQueue<Capacity, ValueType>::popIfSizeIsAtLeast(const uint64_t minSize, ValueType& index) noexcept
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl:235]: Could you please also adjust the name in the definition in the corresponding hpp

File: tools/scripts/used-headers.txt [hunk 17]
```diff
@@ -24,6 +24,7 @@ sys/wait.h
 unistd.h
 # iceoryx_posh / iceoryx_hoofs headers
 algorithm
+array
```
Comment (elBoberido) [inline tools/scripts/used-headers.txt:27]: Would it be much work to revert this change?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 8 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp]: `using Cell = std::atomic<Index>;` is an atomic of lightweight structure, but not a primitive type.

Do we not want to use `std::array` or not? I cannot help but think that our coding or design guidelines combined with the clang-tidy rules are not ideal in this case. We unnecessarily restrict everything and additionally make the code more verbose/ugly.

It is also at odds with the clang tidy rules in this case (which says use something like a std::array). Now we do not have this something of our own yet (and I would argue we do not need it as std::array is fine for us as well).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 12 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp]: Not needed with the current change to std::array, which is under debate.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl (hunk 16 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.inl:235]: Already the case, they were out of sync and that is why it was changed here.

File: tools/scripts/used-headers.txt (hunk 17 above)
Comment (MatthiasKillat) [reply tools/scripts/used-headers.txt:27]: Of course not, but is it necessary? It is currently needed to pass the checks when using `std::array` as substitute for a C array.

It all depends on whether we want/can use `std::array` which I am in favor of.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 12 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp]: @FerdinandSpitzschnueffler As I undid the usage of `std::array` after discussion with @elBoberido I added the justification.

Note that I only use an extra line for the justification if it is too long to fit on the suppression command line. Now we may decide to do it differently in the future but I think this should be a separate PR once the details are decided. (I prefer to have this line optional).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp (hunk 8 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/lockfree_queue/index_queue.hpp:121]: Except that we removed the `std::array` in favor of the C-array to keep the STL dependency low, there is no technical reason not to use it. I would keep the C-array for now and maybe we create a small `cxx::array` wrapper for the `buffer`/`uninitialized_buffer`

**Review by elBoberido — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
